### PR TITLE
Add color scale setting

### LINKED ISSUE #426: Toggling dark mode doesn't change the plotly colour pallete
### Description

Hey there! I'm really loving this project. Previously I was using Weights and Biases, but the network latency was getting to be a real pain so its nice to have something local.

I use dark mode, so the plots look like this:

Screenshot 2023-04-07 at 10 56 56

Which is much less readable than the light mode version:

Screenshot 2023-04-07 at 10 57 07

The dark mode plot is basically impossible to read for me. It would be really helpful if the colour palette was updated when dark mode is selected.

I've got some hours to kill and familiarity with plotly, so am very happy to work on this if someone can point me in the right direction.

### How to Reproduce

1. Start recording data with optuna
1. Open optuna-dashboard
2. Wait for data to come in from the experiments
3. View the Parallel coordinates plot under dark and light mode, and observe the difference.

This is also a problem with other plots, although it is not as pronounced as on the parallel coordinates plot.

### Python version

3.10.5

### Optuna version

3.1.0

### optuna-dashboard version or git revision

0.9.0

### Web browser

Firefox 111.0.1

### PR DESCRIPTION
**Pull request by gen740**
## Contributor License Agreement

This repository (``optuna-dashboard``) and [Goptuna](https://github.com/c-bata/goptuna) share common code.
This pull request may therefore be ported to Goptuna.
Make sure that you understand the consequences concerning licenses and check the box below if you accept the term before creating this pull request.

* [x] I agree this patch may be ported to [Goptuna](https://github.com/c-bata/goptuna) by other Goptuna contributors.

## Reference Issues/PRs

Fixes #426.

## What does this implement/fix? Explain your changes.

- Add the color theme Setting page to the study detail section.
- Add plotly color themes.

### CODE REVIEW

File: optuna_dashboard/index.html
Comment (c-bata) [inline optuna_dashboard/index.html:3]: Could you revert this change since it doesn't related to this PR?

File: optuna_dashboard/ts/components/App.tsx
Comment (c-bata) [inline optuna_dashboard/ts/components/App.tsx:90]: A nit.

```suggestion
```

File: optuna_dashboard/ts/components/Settings.tsx
Comment (HideakiImamura) [inline optuna_dashboard/ts/components/Settings.tsx:71]: How about changing the string color and size as follows?

```suggestion
  return (
    <Grid container spacing={4} sx={{ padding: '40px' }}>
      <Grid item xs={12}>
        <Typography variant="h3" gutterBottom color="textSecondary">
          Settings
        </Typography>
      </Grid>
      <Grid item xs={12}>
        <Typography variant="h5" gutterBottom color="textPrimary">
          Plotly Color Scales
        </Typography>
      </Grid>
      <Grid item xs={2}>
        <Typography variant="h6" color="textSecondary">
          Dark Mode
        </Typography>
      </Grid>
      
      <Grid item xs={10} sx={{ display: 'flex', alignItems: 'center' }}>
        <Select value={darkModeColor} onChange={handleDarkModeColorChange}>
          <MenuItem value={"default"}>Default</MenuItem>
          <MenuItem value={"seaborn"}>Seaborn</MenuItem>
          <MenuItem value={"simpleWhite"}>SimpleWhite</MenuItem>
          <MenuItem value={"presentation"}>Presentation</MenuItem>
          <MenuItem value={"ggplot2"}>GGPlot2</MenuItem>
          <MenuItem value={"gridon"}>GridOn</MenuItem>
          <MenuItem value={"xgridOff"}>XGridOff</MenuItem>
          <MenuItem value={"ygridOff"}>YGridOff</MenuItem>
        </Select>
      </Grid>

      <Grid item xs={2}>
        <Typography variant="h6" color="textSecondary">
          Light Mode
        </Typography>
      </Grid>
      <Grid item xs={10} sx={{ display: 'flex', alignItems: 'center' }}>
        <Select value={lightModeColor} onChange={handleLightModeColorChange}>
          <MenuItem value={"default"}>Default</MenuItem>
          <MenuItem value={"seaborn"}>Seaborn</MenuItem>
          <MenuItem value={"presentation"}>Presentation</MenuItem>
          <MenuItem value={"ggplot2"}>GGPlot2</MenuItem>
          <MenuItem value={"gridon"}>GridOn</MenuItem>
          <MenuItem value={"xgridOff"}>XGridOff</MenuItem>
          <MenuItem value={"ygridOff"}>YGridOff</MenuItem>
        </Select>
      </Grid>
    </Grid>
  )
}
```

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. I have several comments. PTAL.

**Comment by not522** [comment]
@nabenabe0928 Could you review this PR?

**Comment by nabenabe0928** [comment]
> @nabenabe0928 Could you review this PR?

Sure!

File: optuna_dashboard/ts/components/PlotlyDarkMode.ts
Comment (gen740) [reply optuna_dashboard/ts/components/PlotlyDarkMode.ts:18]: Sorry, this is my typo. `Partial<plotly.Template>` is correct.

File: optuna_dashboard/ts/components/Settings.tsx
Comment (gen740) [reply optuna_dashboard/ts/components/Settings.tsx:71]: Seems great!

**Review by HideakiImamura — APPROVED** [review]
LGTM.

As a follow-up, it would be great to save the configuration of the setting page in the some storage to maintain the configuration over the page reload, etc.

**Comment by nabenabe0928** [comment]
Sorry for the late response, but now I am working on the review!

**Review by nabenabe0928 — CHANGES_REQUESTED** [review]
Thank you for the PR!
I am still working on the review, but the first yet most important point is that `GraphRank.tsx` is missing the color scale setting (yes, I know that rank plot was added recently), so could you apply the same modification to it as well?

File: optuna_dashboard/ts/components/AppDrawer.tsx
Comment (nabenabe0928) [inline optuna_dashboard/ts/components/AppDrawer.tsx:314]: Is it better with [ColorLens](https://mui.com/material-ui/material-icons/)?
Optionally, I would like to discuss if "settings" is an appropriate name for this feature because the name is too general.
If we add any further configurations here in the future, I am definitely for it:)

**Review by nabenabe0928 — COMMENTED** [review]
Still work in progress, but checked App.tsx and AppDrawer.tsx.

File: optuna_dashboard/ts/components/Settings.tsx
Comment (nabenabe0928) [inline optuna_dashboard/ts/components/Settings.tsx:57]: I think grid should not be controlled from here.

```suggestion
          <MenuItem value={"default"}>Default</MenuItem>
          <MenuItem value={"seaborn"}>Seaborn</MenuItem>
          <MenuItem value={"presentation"}>Presentation</MenuItem>
          <MenuItem value={"ggplot2"}>GGPlot2</MenuItem>
```

File: optuna_dashboard/ts/components/Settings.tsx
Comment (HideakiImamura) [reply optuna_dashboard/ts/components/Settings.tsx:78]: Sounds good to me.

**Comment by nabenabe0928** [comment]
For future work, it would be helpful to have hints about the color scale in the setting page using [text collapse](https://qiita.com/KensukeTakahara/items/9ba88a9a7a82413b83a1).

**Comment by c-bata** [comment]
@gen740 I'm sorry that I forgot to assign the second reviewer :bow: Could you please resolve the conflict?

**Review by c-bata — CHANGES_REQUESTED** [review]
Thank you for the update. Let me leave an early feedback comment.

File: optuna_dashboard/ts/components/GraphTimeline.tsx
Comment (gen740) [reply optuna_dashboard/ts/components/GraphTimeline.tsx:18]: That change would introduce additional dependencies into `state.ts` (specifically importing `PlotlyColorTemplates.ts`). I believe `state.ts` should remain free of any dependencies. What are your thoughts on this, @c-bata?

Comment (gen740) [reply optuna_dashboard/ts/components/GraphTimeline.tsx:18]: Additionally, `usePlotlyTemplate` cannot be used inside `useEffect` functions, so this change wouldn't simplify the code.

File: optuna_dashboard/ts/components/AppDrawer.tsx
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/AppDrawer.tsx:309]: How about using pop-up to show setting page? I roughly created my example [here](https://github.com/gen740/optuna-dashboard/compare/add_color_scale_setting...keisuke-umezawa:optuna-dashboard:example-popup), although it is very early implementation and needs improvement. 

Some reasons:
1. You can use ${URL_PREFIX}/settings instead of "${URL_PREFIX}/studies/${studyId}/settings" here.
2. You can show change during tweaking theme on settings.

File: optuna_dashboard/ts/components/GraphContour.tsx
Comment (keisuke-umezawa) [inline optuna_dashboard/ts/components/GraphContour.tsx:65]: You need to add `colorTheme` for all useEffect, otherwise the change will not be effected immediately.

File: optuna_dashboard/ts/components/Settings.tsx
Comment (keisuke-umezawa) [inline optuna_dashboard/ts/components/Settings.tsx:54]: Those templates seem to be for Light mode. Should we only provide templates compatible for dark mode?

File: optuna_dashboard/ts/components/AppDrawer.tsx
Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/AppDrawer.tsx:309]: Also, the place of setting should be below because it is for setting for all pages, imo.
スクリーンショット 2024-01-01 11 49 14

You can see this example [here](https://github.com/gen740/optuna-dashboard/compare/add_color_scale_setting...keisuke-umezawa:optuna-dashboard:example-popup)

**Review by keisuke-umezawa — CHANGES_REQUESTED** [review]
I left some review comments. Could you check them?

File: optuna_dashboard/ts/components/Settings.tsx
Comment (gen740) [reply optuna_dashboard/ts/components/Settings.tsx:54]: Unfortunately, there is only one dark template provided in [Plotly's GitHub repository](https://github.com/plotly/plotly.py/blob/v5.6.0/packages/python/plotly/templategen/definitions.py). To offer multiple dark mode templates, we should create original color templates. Therefore, adding (or modifying) an additional dark template could be a suitable follow-up.

Comment (keisuke-umezawa) [reply optuna_dashboard/ts/components/Settings.tsx:54]: Ok. I understand.

**Comment by keisuke-umezawa** [comment]
@gen740 
It looks good to me, and work well in my local env.
In the future, it is better to store those settings somewhere even if we can use the settings when reloading.
Could you make an issue for it to keep it in mind after merging it?

@c-bata @nabenabe0928 
Do you want to check this PR again as you requested changes?

**Review by keisuke-umezawa — APPROVED** [review]

**Comment by gen740** [comment]
I've realized there is a bug that needs to be fixed. Could we please wait to merge this until it's resolved?

**Comment by gen740** [comment]
The bug has been fixed. Could you please take another look? Thank you!

**Comment by c-bata** [comment]
@keisuke-umezawa  I've dismissed reviews from both me and @nabenabe0928 since we've been unassigned from this PR. Please merge this PR after checking the latest changes

**Comment by gen740** [comment]
@keisuke-umezawa
I've resolved the conflicts, so please proceed with the merge. Thank you!

**Pull request merged**
